gcc/rust/util/optional.h

```diff
@@ -1289,24 +1280,28 @@ class optional : private detail::optional_move_assign_base<T>,
   TL_OPTIONAL_11_CONSTEXPR T &value() & {
     if (has_value())
       return this->m_value;
-    throw bad_optional_access();
+
+    gcc_unreachable();
```

Comment:
Just to verify: does `gcc_unreachable` here (and likewise elsewhere) have the correct semantics:
<https://github.com/Rust-GCC/gccrs/blob/7af86eaaffec8cab5ad80e242480fe9f6c1ffaa6/gcc/system.h#L828-L831>
..., that is: an unreachable code path (..., and the compiler should optimized based on that knowledge), or should the original `throw` rather map onto an `abort`:
<https://github.com/Rust-GCC/gccrs/blob/7af86eaaffec8cab5ad80e242480fe9f6c1ffaa6/gcc/system.h#L789-L795>
..., or thus `internal_error`?

Reply:
ah, I know I have used it in other places not as an optimization `unreachable` but more as a "should not reach here `unreachable`", so under the impression that it would always do the `fancy_abort`. I guess we could have a `rust_unreachable`, at least for now, because we're nowhere close to caring about optimization and always want those fancy aborts

Reply:
@tschwinge I've created a branch so the changes are made. I've added it to tomorrow's community call because I want to discuss it with you. I think there is value in having a macro that always `fancy_abort`s and in one that is used for optimization only in releases, but I don't think we are there yet

Reply:
In any way, these changes aren't specific to this PR, so for now I think `tl::optional` should stay similar to the rest of the gccrs codebase and use `gcc_unreachable` (even if I now think it is "invalid" and should be replaced by `fancy_abort`). I will fix it once we've reached a decision on the matter tomorrow :)